neuralop/layers/coda_blocks.py

```diff
@@ -0,0 +1,398 @@
+            fno_skip='linear',
+            norm=None,
+            apply_skip=True,
+            n_layers=1,
+        )
+        self.K = FNOBlocks(
+            output_scaling_factor=1 / scale,
+            conv_module=spectral_convolution,
+            **kqv_args,
+            **common_args,
+        )
+        self.Q = FNOBlocks(
+            output_scaling_factor=1 / scale,
+            conv_module=spectral_convolution,
+            **kqv_args,
+            **common_args,
+        )
+        self.V = FNOBlocks(
+            output_scaling_factor=1,
+            conv_module=spectral_convolution,
+            **kqv_args,
+            **common_args,
+        )
+
+        if self.n_heads * self.head_codimension != self.token_codimension:
+            self.proj = FNOBlocks(
+                in_channels=self.n_heads * self.head_codimension,
+                out_channels=self.token_codimension,
+                n_modes=n_modes,
+                output_scaling_factor=1,
+                # args below are shared with KQV blocks
+                apply_skip=True,
+                non_linearity=torch.nn.Identity(),
+                fno_skip='linear',
+                norm=None,
+                conv_module=spectral_convolution,
+                n_layers=1,
+                **common_args,
+            )
+        else:
+            self.proj = None
+
+        self.attention_normalizer = Normalizer(self.token_codimension)
+
+        mixer_args = dict(
+            n_modes=n_modes,
+            output_scaling_factor=1,
+            non_linearity=non_linearity,
+            norm='instance_norm',
+            fno_skip=fno_skip,
+            conv_module=spectral_convolution,
+        )
+        # We have an option to make the last operator (MLP in regular
+        # Transformer block) permutation equivariant. i.e., applying the
+        # operator per variable or applying the operator on the whole channel
+        # (like regular FNO).
+        if permutation_eq:
+            self.mixer = FNOBlocks(
+                in_channels=self.mixer_token_codimension,
+                out_channels=self.mixer_token_codimension,
+                apply_skip=True,
+                n_layers=2,
+                **mixer_args,
+                **common_args,
+            )
+            self.norm1 = Normalizer(self.token_codimension)
+            self.norm2 = Normalizer(self.mixer_token_codimension)
+            self.mixer_out_normalizer = Normalizer(
+                self.mixer_token_codimension)
+
+        else:
+            self.mixer = FNOBlocks(
+                in_channels=codimension_size,
+                out_channels=codimension_size,
+                n_layers=2,
+                **mixer_args,
+                **common_args,
+            )
+            self.norm1 = Normalizer(codimension_size)
+            self.norm2 = Normalizer(codimension_size)
+            self.mixer_out_normalizer = Normalizer(codimension_size)
+
+    def compute_attention(self, xa, batch_size):
```

Comment:
Let's try to give helpful names to the variables - `xa` isn't very clear

Reply:
why not `tokens`?